OpenRA.Mods.RA2/Traits/AttackGarrisonedRV.cs

```diff
@@ -0,0 +1,221 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2018 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Mods.Common.Traits.Render;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.RA2.Traits
+{
+	public class FirePort
+	{
+		public WVec Offset;
+		public WAngle Yaw;
+		public WAngle Cone;
+	}
+
+	[Desc("Garrisoners can fire their weapons out of fire ports.")]
+	public class AttackGarrisonedRVInfo : AttackFollowInfo, IRulesetLoaded, Requires<GarrisonableInfo>
```

Comment:
Please don't duplicate traits. Submit your changes to first instead: https://github.com/OpenRA/OpenRA/blob/bleed/OpenRA.Mods.Common/Traits/Attack/AttackGarrisoned.cs

Reply:
I have no idea how much impact replacing CargoInfo with GarrisonableInfo would cause here.

Reply:
I see. https://gist.github.com/Mailaender/accd6464cb1b7a12ccad5d75256f1da2/revisions In that case the whole pull request should probably move to https://github.com/OpenRA/OpenRA.

Reply:
I think they refused the whole PR already...

Reply:
Hmm... Duplication isn't good. Rename to `AttackFromGarrison` or something then at least.

Reply:
Surely I can rename it but I just wanted to let you know I might not have the engine knowledge to do any better.

Reply:
I also can't think of a clever way of using inheritance. This might be the only way to go.

Reply:
What's next then ? What would you like to do and how ?

Reply:
This pull request seems to differ from what was initially submitted at https://github.com/OpenRA/OpenRA/pull/12774/ so I suggest a new attempt at the engine project. You can rebase a copy to `playtest-20200118` type in the commit into https://github.com/OpenRA/ra2/blob/master/mod.config#L7 and keep a test case with the .yaml changes here. See some of my pull requests with `Engine Update Required` as reference.